src/particles/ImpactXParticleContainer.cpp

```diff
@@ -111,19 +110,22 @@ namespace impactx
          * is then copied to the permanent tile which is stored on the particle
          * (particle_tile).
          */
-        using PinnedTile = amrex::ParticleTile<NStructReal, NStructInt, NArrayReal, NArrayInt,
+        using PinnedTile = amrex::ParticleTile<amrex::SoAParticle<RealSoA::nattribs, IntSoA::nattribs>, RealSoA::nattribs, IntSoA::nattribs,
                 amrex::PinnedArenaAllocator>;
+        using SoAParticle = amrex::SoAParticle<RealSoA::nattribs, IntSoA::nattribs>;
+
         PinnedTile pinned_tile;
+        auto ptd = pinned_tile.getParticleTileData();
         pinned_tile.define(NumRuntimeRealComps(), NumRuntimeIntComps());
```

Comment:
It looks like `define()` is not allocating the former AoS attributes that we access in the next lines via `PTD`.

Check what `ParticleTile::define` is calling and what `StructOfArrays::define` manages - I think `m_rdata` and `m_idata` are not yet allocated.

Reply:
Rough idea:
```C++
{
        BL_PROFILE("ImpactX::AddNParticles");

        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(lev == 0, "AddNParticles: only lev=0 is supported yet.");
        AMREX_ALWAYS_ASSERT(x.size() == y.size());
        AMREX_ALWAYS_ASSERT(x.size() == z.size());
        AMREX_ALWAYS_ASSERT(x.size() == px.size());
        AMREX_ALWAYS_ASSERT(x.size() == py.size());
        AMREX_ALWAYS_ASSERT(x.size() == pz.size());

        // number of particles to add
        int const np = x.size();

        // have to resize here, not in the constructor because grids have not
        // been built when constructor was called.
        reserveData();
        resizeData();

        // write Real attributes (SoA) to particle initialized zero
        auto& particle_tile = DefineAndReturnParticleTile(0, 0, 0);

        /* Create a temporary tile to obtain data from simulation. This data
         * is then copied to the permanent tile which is stored on the particle
         * (particle_tile).
         */
        using PinnedTile = amrex::ParticleTile<amrex::SoAParticle<RealSoA::nattribs, IntSoA::nattribs>, RealSoA::nattribs, IntSoA::nattribs,
                amrex::PinnedArenaAllocator>;

        PinnedTile pinned_tile;
        pinned_tile.define(NumRuntimeRealComps(), NumRuntimeIntComps());
        pinned_tile.resize(np);

        for (int i = 0; i < np; i++)
        {
            pinned_tile.id(i) = amrex::SoAParticle<RealSoA::nattribs, IntSoA::nattribs>::NextID();
            pinned_tile.cpu(i) = amrex::ParallelDescriptor::MyProc();
            pinned_tile.pos(i, 0) = x[i];
            pinned_tile.pos(i, 1) = y[i];
            pinned_tile.pos(i, 2) = z[i];

            pinned_tile.getParticleTileData().m_rdata[RealSoA::ux][i] = px[i];
            pinned_tile.getParticleTileData().m_rdata[RealSoA::uy][i] = py[i];
            pinned_tile.getParticleTileData().m_rdata[RealSoA::pt][i] = pz[i];
            pinned_tile.getParticleTileData().m_rdata[RealSoA::m_qm][i] = qm;
            pinned_tile.getParticleTileData().m_rdata[RealSoA::w][i] = bchchg/ablastr::constant::SI::q_e/np;
        }

        /* Redistributes particles to their respective tiles (spatial bucket
         * sort per box over MPI ranks)
         */
        auto old_np = particle_tile.numParticles();
        auto new_np = old_np + pinned_tile.numParticles();
        particle_tile.resize(new_np);
        amrex::copyParticles(
                particle_tile, pinned_tile, 0, old_np, pinned_tile.numParticles());
    }
```

but something is still invalid in this, because `RedistributeCPU` still fails when doing this.

Reply:
Ah no wrong, the other way around.
We can ignore the first `for` loop and just do `pinned_tile.push_back_real(RealSoA::ux, px);` for every property now.

Reply:
```C++
{
        BL_PROFILE("ImpactX::AddNParticles");

        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(lev == 0, "AddNParticles: only lev=0 is supported yet.");
        AMREX_ALWAYS_ASSERT(x.size() == y.size());
        AMREX_ALWAYS_ASSERT(x.size() == z.size());
        AMREX_ALWAYS_ASSERT(x.size() == px.size());
        AMREX_ALWAYS_ASSERT(x.size() == py.size());
        AMREX_ALWAYS_ASSERT(x.size() == pz.size());

        // number of particles to add
        int const np = x.size();

        // have to resize here, not in the constructor because grids have not
        // been built when constructor was called.
        reserveData();
        resizeData();

        auto& particle_tile = DefineAndReturnParticleTile(0, 0, 0);

        /* Create a temporary tile to obtain data from simulation. This data
         * is then copied to the permanent tile which is stored on the particle
         * (particle_tile).
         */
        using PinnedTile = amrex::ParticleTile<amrex::SoAParticle<RealSoA::nattribs, IntSoA::nattribs>, RealSoA::nattribs, IntSoA::nattribs,
                                               amrex::PinnedArenaAllocator>;

        PinnedTile pinned_tile;
        pinned_tile.define(NumRuntimeRealComps(), NumRuntimeIntComps());

        for (int i = 0; i < np; i++)
        {
            pinned_tile.push_back_int(IntSoA::id, ParticleType::NextID());
            pinned_tile.push_back_int(IntSoA::cpu, amrex::ParallelDescriptor::MyProc());
        }

        // write Real attributes (SoA) to particle initialized zero
        DefineAndReturnParticleTile(0, 0, 0);

        pinned_tile.push_back_real(RealSoA::x, x);
        pinned_tile.push_back_real(RealSoA::y, y);
        pinned_tile.push_back_real(RealSoA::z, z);
        pinned_tile.push_back_real(RealSoA::ux, px);
        pinned_tile.push_back_real(RealSoA::uy, py);
        pinned_tile.push_back_real(RealSoA::pt, pz);
        pinned_tile.push_back_real(RealSoA::m_qm, np, qm);
        pinned_tile.push_back_real(RealSoA::w, np, bchchg/ablastr::constant::SI::q_e/np);

        /* Redistributes particles to their respective tiles (spatial bucket
         * sort per box over MPI ranks)
         */
        auto old_np = particle_tile.numParticles();
        auto new_np = old_np + pinned_tile.numParticles();
        particle_tile.resize(new_np);
        amrex::copyParticles(
                particle_tile, pinned_tile, 0, old_np, pinned_tile.numParticles());
    }
```

Reply:
pushed :)